### ISSUE TITLE
Better error messages for the monomorphism restriction

### ISSUE THREAD

**Issue by noughtmare** [haskell/error-messages#40]
Today [Competitive_Ad2539 on Reddit asks why type inference fails for this code](https://reddit.com/r/haskellquestions/comments/sp9u4a/why_did_the_type_inference_failed_here/):

```haskell
showLine = (++ "\n") . concat
```

The error message is (with GHC 9.2.1 and 9.0.2):

```
T.hs:1:24: error:
    * Ambiguous type variable `t0' arising from a use of `concat'
      prevents the constraint `(Foldable t0)' from being solved.
      Relevant bindings include
        showLine :: t0 [Char] -> [Char] (bound at T.hs:1:1)
      Probable fix: use a type annotation to specify what `t0' should be.
      These potential instances exist:
        instance Foldable (Either a) -- Defined in `Data.Foldable'
        instance Foldable Maybe -- Defined in `Data.Foldable'
        instance Foldable ((,) a) -- Defined in `Data.Foldable'
        ...plus two others
        ...plus 26 instances involving out-of-scope types
        (use -fprint-potential-instances to see them all)
    * In the second argument of `(.)', namely `concat'
      In the expression: (++ "\n") . concat
      In an equation for `showLine': showLine = (++ "\n") . concat
  |
1 | showLine = (++ "\n") . concat
  |                        ^^^^^^
```

Experts might see at a glance that this is probably caused by the monomorphism restriction, but that is not at all obvious from this error message.

I would propose adding a note in the error message that the monomorphism restriction prevents it from inferring the most general type.

**Comment by goldfirere** [comment]
I don't think the MR would work as a post-pass. For example:

```hs
x = 5

y :: Int
y = x

z :: Double
z = x
```

If we infer a most general type for `x`, then the rest of the program is accepted. But now how should the post-pass work? Maybe we can accumulate (off to the side somewhere) all the places where `x` is instantiated, and then make sure there's a possible solution, but that would be quite an elaborate system.

On the other hand, we *could* easily remember somewhere what the most general type of a MR'd variable would be and include that in error messages. That would probably help a good deal.

**Comment by Ericson2314** [comment]
Ah, good points. Is the most general type of the MR'd variable locally definable?

Maybe we should just....type check both ways, and at the later point when our type checker is not so massive because it uses more abstractions, find a way to deduplicate those a bit.

(The same underlying mechanism would be good for the CPP check-all-combinations @mpickering and I talked about, FWIW.)

**Comment by goldfirere** [comment]
> Ah, good points. Is the most general type of the MR'd variable locally definable?

Yes. The MR is implemented in one specific place (`GHC.Tc.Solver.decideMonoTyVars`). Though there would be corner cases that could go awry, in most cases, it should be easy enough to produce both an MR'd type and a non-MR'd type. This would be more than just changing error text, though, and so it would be motivating to have a number of examples of error messages without this feature and with this feature, to see how concretely it would improve things.

**Comment by Ericson2314** [comment]
OK so the corner case would be something like mutually recursive bindings subject to the MR influence type inference of each other, right?

**Comment by goldfirere** [comment]
It's hard for me to say exactly without taking an honest stab at the implementation. This stuff is surprisingly tricky.
